Feature/myf 55 view form template for signup

kelly-mcdougall opened the pull request:
## This pull request contributes to MYF-57, MYF-55, and MYF-17:

### What is included in this pull request
- adds basic registration page
- adds basic form functionality to allow a user to create a new account (username, password, email)

### How to test
- Launch the application, either with docker or on local machine
- Go to localhost:8000/register in your browser
- You should see a basic registration screen
- Fill out the info and you should be redirected to the homepage.

### Remaining work
- I still need to stub out a new page to redirect users
- additional form validation needed
- incorporate groups as part of signup process
- At the moment users can only register, I need to repurpose view to also allow sign-in functionality if the user has already registered.

chrxr committed:
adding authenticate app to house sign-up and login/out views, forms and templates

kelly-mcdougall committed:
- Contributes to MYF-57, MYF-55, and MYF-17.
- Updates authentication app
- adds form functionality to allow signups

chrxr commented on app/app/urls.py:
```diff
@@ -14,8 +14,12 @@
 urlpatterns = [
     path('admin/', admin.site.urls),
+    path('', include(homepage_urls)),
+    path('register', include(authenticate_urls)),
```
URLs should go from most specific down to least specific, so I would suggest moving 'register' up one so that the root is at the bottom. Also, without the slash at the end of register, `register/` will result in a 404. If you add the `/` it matches both `register/` and `register`.

chrxr commented on app/authenticate/forms.py:
```diff
@@ -0,0 +1,15 @@
+
+
+class SignUpForm(UserCreationForm):
+    username = forms.CharField(max_length=30, required=False, help_text='Optional.', widget=forms.TextInput(attrs={'placeholder': 'Username', 'class': "form-control"}))
```
Username is listed as optional, however, it is used explicitly in the `user_register` view. Utilizing email for login will be a little more complicated to implement, so I would suggest that we just make username a required field.

chrxr commented on app/authenticate/views.py:
```diff
@@ -0,0 +1,23 @@
+        form = SignUpForm(request.POST)
+        if form.is_valid():
+            form.save()
+            username = form.cleaned_data.get('username')
```
As noted above, this can break because username is not a required field. Making this required in the form will fix this issue.

chrxr commented on app/authenticate/forms.py:
```diff
@@ -0,0 +1,15 @@
+from django import forms
+from django.contrib.auth.forms import UserCreationForm
+from django.contrib.auth.models import User
```
This is a minor point, but it was raised by py-linter when I ran it against the code. A better way to reference the user model is below:
```suggestion
from django.contrib.auth import get_user_model
```

then use that function below:

```
    class Meta:
        model = get_user_model()
```

The reason for this is that `get_user_model()` gets the actual user model as set in `app.settings` or the default `user` model if a custom model is not being used. So if you wanted to use a custom `user` model in the future, you wouldn't need to alter any of the imports in other files.

chrxr commented on app/authenticate/views.py:
```diff
@@ -0,0 +1,23 @@
+from django.shortcuts import render, redirect
+from django.contrib.auth.models import Group
+from django.contrib.auth import login, logout, authenticate
```
Logout is an unused import

chrxr commented on app/authenticate/views.py:
```diff
@@ -0,0 +1,23 @@
+from django.contrib.auth.models import Group
+from django.contrib.auth import login, logout, authenticate
+# from django.contrib.auth.forms import UserCreationForm
+from django.urls import reverse_lazy
+from django.views import generic
```
Unused imports

chrxr commented on app/authenticate/views.py:
```diff
@@ -0,0 +1,23 @@
+            raw_password = form.cleaned_data.get('password1')
+            user = authenticate(username=username, password=raw_password)
+            login(request, user)
+            return render(request, 'homepage/homepage.html')
```
Rather than using `render`, should we use `redirect` instead? For example:

```
from homepage.views import homepage

def user_register(request):
    ...
    return redirect(homepage)
```

With render, the URL in the address bar remains at `/register/`, whereas with redirect it triggers a new request for the homepage at the site root.

In the future, we may want to include a 'next' attribute which remember which page the user came from then redirects them back to there. Something to think about in the future.

chrxr commented on template.env:
```diff
@@ -8,7 +8,7 @@
 
 # A list of domains and/or IPs that Django is allowed to serve
-DJANGO_ALLOWED_HOSTS=localhost 127:0.0.1 [::1]
+DJANGO_ALLOWED_HOSTS="localhost 127:0.0.1 [::1]"
```
I don't believe quoting this string should be necessary.

chrxr commented on docker-compose.prod.yml:
```diff
@@ -4,7 +4,7 @@ services:
     build:
       context: ./app
-      dockerfile: Dockerfile
+      dockerfile: app/Dockerfile
```
Did you need to change this to get this docker file to work? The `dockerfile` setting is a relative path from the `context` setting above, and so it should work without `app/`. I'd actually suggest not using the `docker-compose.prod` file locally, and instead just use `docker-compose up`. We can chat about this on our next call if need be.

chrxr commented on app/authenticate/templates/authenticate/authenticate.html:
```diff
@@ -0,0 +1,23 @@
+
+    <h1>register <br></h1>
+
+    <h5>Please sign up</h5> {% for field in form %}
```
Could we format these `{%for ..%}` blocks a little differently so it's easier to see the nesting of the blocks? Perhaps just more this first block to below the `<h5>`, then break the two `{% endfor %}` tags on line 17 on to two lines.

chrxr reviewed (changes requested):
Hi @kelly-mcdougall. Looks good! Just a few small suggestions. Let’s leave this PR open and add tests to it over the next week.

kelly-mcdougall committed:
Correcting registration page

- housekeeping to remove unused imports
- updated authenitcation.views to use redirect to return to homepage after registration
-reordered URLs in app/urls.py

chrxr commented on app/authenticate/forms.py:
```diff
@@ -0,0 +1,15 @@
+
+
+class SignUpForm(UserCreationForm):
+    username = forms.CharField(max_length=30, required=True, help_text='Optional.', widget=forms.TextInput(attrs={'placeholder': 'Username', 'class': "form-control"}))
```
`help_text` for this should be "Required"

chrxr committed:
changing help text on form for username to 'required'

chrxr reviewed (approved):
Hi @kelly-mcdougall, this looks great! Thanks. I pushed one tiny change, changing the help text for the username field on the form to "required". Let's get this merged!
